src/main.cpp

```diff
@@ -996,6 +997,51 @@ bool ContextualCheckTransaction(
                             REJECT_INVALID, "bad-txns-oversize");
     }
 
+    // Rules that apply before Heartwood:
+    if (!heartwoodActive) {
+        if (tx.IsCoinBase()) {
+            // A coinbase transaction cannot have output descriptions
+            if (tx.vShieldedOutput.size() > 0)
+                return state.DoS(100, error("CheckTransaction(): coinbase has output descriptions"),
+                                REJECT_INVALID, "bad-cb-has-output-description");
+        }
+    }
+
+    // Rules that apply to Heartwood or later:
+    if (heartwoodActive) {
+        if (tx.IsCoinBase()) {
+            // All Sapling outputs in coinbase transactions MUST have valid note commitments
+            // when recovered using a 32-byte array of zeroes as the outgoing viewing key.
+            uint256 ovk;
```

Comment:
That makes things less readable, I think. I can't call it directly with `0` as `uint256` isn't a std-lib array type, so it would instead be using the `uint256(const std::vector<unsigned char>&)` initializer which would be much less intuitive. I could call `ovk.SetNull()` explicitly afterwards if you want, but that just results in `memset(data, 0, sizeof(data))` being called twice (once each for `uint256 ovk` and `ovk.SetNull()`).

Reply:
Agreed.